Tools/muscan.1
.TH "MUSCAN" "1" "Release 0.1.13" "daelstorm" "Museek Daemon Plus"
.SH "NAME"
.LP 
\fBMuscan\fR \- File scanner for Museek
.SH "SYNOPSIS"
.B muscan
[\-c <\fIfilename\fP>] [\-\-config <\fIfilename\fP>]
[\-b] [\-\-buddy]
[\-v] [\-\-verbose]
[\-r] [\-\-rescan]
[\-n] [\-\-noscan]
[\-s <\fIdirectory\fP>] [\-\-share <\fIdirectory\fP>]
[\-u <\fIdirectory\fP>] [\-\-unshare <\fIdirectory\fP>]
[\-l] [\-\-list]
[\-h] [\-\-help]
.SH "DESCRIPTION"
.LP 
Muscan scans paths & files to be shared by \fImuseekd\fP(1). It creates a database of files, with meta\-data for MP3 and OGG Vorbis files, if Vorbis support was compiled in. There are two available shares databases, Normal and Buddies\-Only. Buddies\-Only is an optional shares database that, if enabled, is only accessable by the users you've chosen as "Buddies".
.LP 
Before running muscan, you will need a configured museekd,  which can be done with \fImusetup\fP(1) or fI>musetup\-gtk\fP(1), and add some shared paths with either of the setup tools or with \fImuscan\fP(1).
.SH "OPTIONS"
.LP 
Muscan accepts the following options:
.TP 
.B \-c <\fIfilename\fP>, \-\-config <\fIfilename\fP>
Use a different config file.
.TP 
.B \-b, \-\-buddy
Select Buddy\-Only Shares.
.TP 
.B \-s <\fIdirectory\fP>, \-\-share <\fIdirectory\fP> 
Add directory to shares.
.TP 
.B \-u <\fIdirectory\fP>, \-\-unshare <\fIdirectory\fP> 
Remove directory from shares.
.TP 
.B \-r, \-\-rescan
Rescan shares.
.TP 
.B \-n, \-\-noscan
Do not rescan shares.
.TP 
.B \-v, \-\-verbose
Be verbose while scanning shares.
.TP 
.B \-l, \-\-list
Display list of shared directories.
.TP 
.B \-h, \-\-help
Display Help, version and exit.
.SH "EXAMPLE"
.LP 
To run a quick shares update, run:
.LP 
\fBmuscan\fP or \fBmuscan \-b\fP
.LP 
To rescan everything, run:
.LP 
\fBmuscan \-r\fP or \fBmuscan \-b \-r\fP
.LP 
Alternatively you use a different config file with:
.LP 
\fBmuscan \-\-config config.xml\fP
.LP 

.SH "FILES"
.TP 
 \fI~/.museekd/config.xml\fR
The default location for the \fBmuseekd\fP config file.
.TP 
 \fI~/.museekd/config.shares\fR
The default location for the Normal shares database.
.TP 
 \fI~/.museekd/config.shares.state\fR
The default location for the active Normal shares database.
.TP 
 \fI~/.museekd/config.buddyshares\fR
The default location for the Buddy shares database.
.TP 
 \fI~/.museekd/config.buddyshares.state\fR
The default location for the active Buddy shares database's.
.SH "AUTHORS"
.LP 
Hyriand <hyriand@thegraveyard.org>
.LP 
daelstorm <daelstorm@gmail.com>
.SH "SEE ALSO"
.LP 
\fImucous\fP(1) \fImulog\fP(1) \fImurmur\fP(1) \fImuscand\fP(1) \fImuseekcontrol\fP(1) \fImuseekd\fP(1) \fImuseeq\fP(1) \fImusetup\fP(1) \fImusetup\-gtk\fP(1)


CHANGELOG
Museek Plus Changelog

0.1.13:  August 25th 2007

BUILD SYSTEMS
CMake, qmake, SCons and Python's distutils
 * RECOMMENDED: cmake
 * CMake compiles and installs everything. Read the INSTALL files for details.
 * qmake compiles just museeq. Read museeq/INSTALL
 * You can install the python bindings, apps, and clients with distutils.
   Run "python setup.py install" in each python sub-directory
 * SCons is still available, but not recommended. Read INSTALL.scons

MUSEEK DAEMON
 * Ignore SIGPIPE (to prevent crashes on BSDs)
 * Catch std::bad_alloc when unpacking a network message.
 * Always get IP from server, when requested
 * Autoretry downloads
 * Syslog support
 * Watch trusted users' status

MUSEEQ
 * Settings revamped
 * Shares configuration added to Settings
 * Added Audacioous to NowPlaying QSA script
 * Added colors to transfers' status column

MUCOUS
 * Setup revamped
 * Many, many bugfixes

MUSEEKCONTROL
 * Implemented a few more unfinished options

MURMUR
 * The basic features are here, enough for inclusion in this release.

0.1.12:  October 21th 2006

MUSEEK DAEMON
 * Fixed museekd not reacting on connect mode changes during run-time (hyriand)
 * Fixed the clear-initiating-transfer that causes a delayed crash (hyriand)
 * Character unpacking checks (fixes corrupted message crashes) (hyriand)
 * Trusted Users' status is now watched
 * IMPORTANT NOTE: the CFLAGS variable renamed to FLAGS and is now a String 
   instead of List, your mulocal.py will need to be modified or erased to allow
   SCons to run correctly. Not doing so will result in SCons failing at the 
   first compile-test.
 * Change the way the IP Address string is created to fix endian bug (hyriand) 
 * New SCons variable LOCKFLAGS disables SCons' automatic flag adding
 * Replaced depreciated call to Send Speed (34) on download completion
   with Send Upload Speed (121) on upload completion. (daelstorm)
 * Check for read/write permissions to config file before loading (daelstorm)
 * Added Global message (66) handler, and pass it to the clients (daelstorm)
 * Removed disconnect-on-hangup-signal (caused a loop)
 * Client message 'Download file to a directory' added (daelstorm)
 * Sends a Museek Daemon version message to clients on connection (daelstorm)
 * Sends the number of privileged users to clients on Server Login (daelstorm)

MUSCAN
 * Ignore .dotfiles and .dotdirs/
 * Removing trailing /s for shared directories
 
MUSEEQ
 * Fixed crash when disconnecting while still connecting to museekd by using
   "mSocket->deleteLater()" instead of "delete mSocket" (hyriand)
 * Fixed method of creating UNIX socket (hyriand)
 * Check option in settings to toggle display of exit dialog
 * Added nick's transfer list color patch transfers are now different colors,
   depending upon their state.
 * Added MUSEEQTRAYICON scons option to enable/disable trayicon at buildtime.
 * Moved Trayicon into Museeq class, made it change with away status and added
   menu option to toggle it.
 * Added /commands to private chat
 * Autoconnect to daemon option added
 * IP addresses can now be viewed in the log window
 * Updated Italian translation (requiem)
 * Updated Polish translation (belegdol)
 
MUSETUP-GTK
 * Scons tries more than one method to detect PyGTK
 
MUCOUS
 * The Setup Interface is now completely controllable from the keyboard
 * Browse Shares now has expandable/collapsable directories and resizeable
   windows. There are some other UI improvements, as well.
 * Added /me and status colors to Private chat 
 * Rewrote tab completion
 * Case-insensitive sorting
 * Refactored / Renamed lots of functions
 * Killed dozens of memory leaks
 * Removed vertical borders from private chat and rooms
 * Added scrolling ticker to chat rooms
 * Improved the Transfers colors and use vertical lines instead of pipes
 * Fixed some broken config options
 * Simplified creation of setup widgets
 * Setup's control over muscan is slightly better, attempts to detect museekd
   config, to better scan correct shares databases.
 * Hundreds of minor changes
 
0.1.11:  July 7th 2006

MUSEEK DAEMON
* ReAdd the server ping, but change the delay between them from 15 to 45 seconds
  to reduce load on the server and to not get banned :).
* Try to logout when shutting down
* Fixed the minPort == maxPort bug and maxPort > minPort bug, for peer
  connections. Previously, it'd skip the port listen attempt entirely. Now it
  tries the listen on minPort (if they equal) or on maxPort (if it's smaller 
  than minport) and  then gives up if it fails. 
* SCons can now set CCFLAGS with the CFLAGS list option
* Enable or disable installation of musetup-gtk with Scons bool option:
  MUSETUPGTK
* Choose which translations to install with the MUSEEQTRANSLATIONS Scons list
  option. (New translations may require modification to the SConstruct file)

MUSEEQ
* Added Timestamp toggle to the Settings menu
* Made user status appear in Private Chat (* user is away || * user is offline )
* Ported all .ui dialogs to cpp
* Added encodings comboboxes to settings dialog, as well as making the password
  dialog protected with ****s
* Added Translation *Support*
* Le_Vert translated museeq into French
* Requiem translated museeq into Italian
* Belegdol translated museeq into Polish
* SyphiliticVagina translated museeq into Spanish
* () translated museeq into German
* Senhor_das_Moscas translated museeq into Brazilian Portuguese
* Added more icons (ip, comments, alert, and small-icons for:  browse, userinfo,
  search, chatrooms, privatechat )
* Add two of mike-labo's themes from museeq. Read the README files for credits.

MUCOUS
* Version to 0.8.2
* New commands: /autoclear, /autoretry, /beep, /trustlist
* Add trusted users list to Users mode
* Make mucous behave more nicely when not connected to a museekd.
* Reorganize parts of setup mode, add new options for buddy-only shares, etc.
* Add timestamps to Debug log, bugfixes for browse mode, topbar variables
  merged into the new storage method


murmur/pymurmur/muscan.py
# This is part of the Murmur Museek Client, and distributed under the GPLv2
# Copyright (c) 2007 daelstorm.
import threading
import os

## Muscan commands, threads
#
class Muscan:
	## Constructor
	# Launch a local muscan process with special options from the setup buttons,
	# Or from /commands
	## @author daelstorm
	# @param self Muscan (Class)
	# @param frame Murmur (Class)
	def __init__(self, frame):
		## @var frame
		# Murmur (Class)
		self.frame = frame
		## @var timer
		# threading timer that calls ThreadMuscan
		self.timer = threading.Timer(1.0, self.ThreadMuscan)
		## @var command
		# once set, a list of command options for subprocess
		self.command = []
		self.configfile = None
		
	def GetConfig(self):
		self.configfile = None
		if self.frame.Networking.config["shares"]["database"] != "":
			fileprefix = self.frame.Networking.config["shares"]["database"].rsplit(".", 1)
			if len(fileprefix) == 2:
				file = fileprefix[0]+".xml"
				if os.path.exists(file):
					self.configfile = file
					
	## Clear timer and restart it in one second
	# @param self Muscan (Class)
	def RestartTimer(self):
		
		self.timer.cancel()
		if self.configfile != None:
			self.timer = threading.Timer(1.0, self.ThreadMuscan)
			self.timer.start()

	## Set command and call RestartTimer
	# @param self Muscan (Class)
	# @param command List of command options for subprocess
	def Command(self, command):
		self.command = command
		self.RestartTimer()

	## Spawn Subprocess and accept stdout and stderr
	# @param self Muscan (Class)
	def ThreadMuscan(self):
		try:
			self.timer.cancel()
			if self.frame.subprocess_fail:
				self.frame.AppendToLogWindow("This feature requires Python 2.4")
				return
			if self.configfile == None:
				self.frame.AppendToLogWindow("Museekd config file not found")
				return
			if self.frame.Config["connection"]["interface"][:9] in ("localhost", "/tmp/muse") and self.command != [] :
				p = "/usr/bin/muscan"
				if os.path.exists(p):
				
					z = self.frame.subprocess.Popen( self.command, stdout=self.frame.subprocess.PIPE, stderr=self.frame.subprocess.PIPE)
					stdout_text, stderr_text = z.communicate()
					z.wait()
					
					stdout_text = stdout_text.split('\n')
					stderr_text = stderr_text.split('\n')
					for line in stdout_text:
						if line.isspace() or line == '':
							pass
						else:
							self.frame.AppendToLogWindow(line)
					for line in stderr_text:
						if line.isspace() or line == '':
							pass
						else:
							self.frame.AppendToLogWindow(line)
					self.frame.AppendToLogWindow("Finished with shares.")
			else:
				self.frame.AppendToLogWindow("Your Museekd is either running remotely or already running a command, cancelling.")
		
			self.command = []
		except Exception,e:
			self.frame.AppendToLogWindow("Error: ThreadMuscan: " + str(e), 1)

	## Change setup input mode, so directory paths can be inputted
	# @param self Muscan (Class)
	# @param inputmode string that is set as Spl["setup_input"] variable
	def ChangeInput(self, inputmode):
		self.frame.Setup.input = inputmode
		self.frame.Setup.Mode()

	## Output list of Normal shared directories to debug log
	# @param self Muscan (Class)
	def ListNormal(self):
		self.command = ["muscan", "-c", self.configfile, "-l"]
		self.RestartTimer()
		self.frame.AppendToLogWindow("Listing normal shares with muscan:")

	## Output list of Buddy-only shared directories to debug log
	# @param self Muscan (Class)
	def ListBuddy(self):
		self.command = ["muscan", "-c", self.configfile, "-b", "-l"]
		self.RestartTimer()
		self.frame.AppendToLogWindow("Listing buddy shares with muscan:")

	## Rescan Buddy-only Shares (rebuilds shares from scratch)
	# @param self Muscan (Class)
	def RescanBuddy(self):
		self.command = ["muscan", "-c", self.configfile, "-v", "-b", "-r"]
		self.RestartTimer()
		self.frame.AppendToLogWindow("Rescanning buddy shares with muscan, don't forget to Reload them. Please wait for muscan to complete.")

	## Update Buddy-only Shares (checks mtimes and updates only newer dirs/files)
	# @param self Muscan (Class)
	def UpdateBuddy(self):
		self.command = ["muscan", "-c", self.configfile, "-v", "-b"]
		self.RestartTimer()
		self.frame.AppendToLogWindow("Updating buddy shares with muscan, don't forget to Reload them. Please wait for muscan to complete.")

	## Update Normal Shares (checks mtimes and updates only new dirs) 
	# @param self Muscan (Class)
	def UpdateNormal(self):
		self.command = ["muscan", "-c", self.configfile, "-v"]
		self.RestartTimer()
		self.frame.AppendToLogWindow("Updating shares with muscan, don't forget to Reload them. Please wait for muscan to complete.")

	## Rescan Normal Shares (rebuilds shares from scratch)
	# @param self Muscan (Class)
	def RescanNormal(self):
		self.command = ["muscan", "-c", self.configfile, "-v", "-r"]
		self.RestartTimer()
		self.frame.AppendToLogWindow("Rescanning shares with muscan, don't forget to Reload them. Please wait for muscan to complete.")
		
	def AddNormalDirectory(self, directory):
		dir = os.path.expanduser(directory)
		if os.path.exists(dir):
			
			self.command = ["muscan", "-c", self.configfile, "-v", "-s", dir]
			self.frame.AppendToLogWindow( "Adding %s to normal shares. Scanning will begin." % dir)
			self.RestartTimer()
		else:
			self.frame.AppendToLogWindow( "Warning: Directory does not exist: %s" % dir)
	def RemoveNormalDirectory(self, directory):
		dir = os.path.expanduser(directory)
		self.command = ["muscan", "-c", self.configfile, "-v", "-u", dir]
		self.frame.AppendToLogWindow( "Removing %s from normal shares. Please rescan or update." % dir)
		self.RestartTimer()
			
	def AddBuddyDirectory(self, directory):
		dir = os.path.expanduser(directory)
		if os.path.exists(dir):
			self.command = ["muscan", "-c", self.configfile, "-v", "-b", "-s", os.path.expanduser(directory)]
			self.frame.AppendToLogWindow( "Adding %s to buddy shares. Scanning will begin." % dir)
			self.RestartTimer()
		else:
			self.frame.AppendToLogWindow( "Warning: Directory does not exist: %s" % dir)
	def RemoveBuddyDirectory(self, directory):

		dir = os.path.expanduser(directory)
		self.command = ["muscan", "-c", self.configfile, "-v", "-b", "-u", dir]
		self.frame.AppendToLogWindow( "Removing %s from buddy shares. Please rescan or update." % dir)
		self.RestartTimer()